service/add_citation.py

```diff
@@ -2,83 +2,88 @@
+    if paper_id not in cited_paper_id_to_cnt:
+        cited_paper_id_to_cnt[paper_id] = cite_cnt
+        cited_paper_list.append((paper_id, paper_title))
+        cite_cnt += 1
+    return cited_paper_id_to_cnt[paper_id], cite_cnt
+
+def create_cite_string(paper_id, cite_cnt):
+    """Create a citation string for a paper."""
+    return f"[[{cite_cnt}]](https://arxiv.org/abs/{paper_id})"
+
+def add_citation_to_response(final_response, right, cite_str):
+    """Add a citation string to the response."""
+    return final_response[:right - 1] + cite_str + final_response[right - 1:]
+
+def create_references_list(cited_paper_list):
+    """Create a formatted string of references."""
+    cited_list_str = ""
+    for cite_idx, (cited_paper_id, cited_paper_title) in enumerate(cited_paper_list, start=1):
+        cited_list_str += f"""[[{cite_idx}] {cited_paper_title}](https://arxiv.org/abs/{cited_paper_id})\n\n"""
+    return cited_list_str
+
 def add_citation_with_retrieved_node(retrieved_nodes, final_response):
-    if retrieved_nodes is None or len(retrieved_nodes) <= 0:
+    """Main function to add citations to the response."""
+    if not retrieved_nodes:
         return final_response
-    bm25_retriever = BM25Retriever.from_defaults(nodes=retrieved_nodes, similarity_top_k=2)
+
+    bm25_retriever = create_bm25_retriever(retrieved_nodes)
     sentences = [sentence for sentence in split_sentences(final_response) if len(sentence) > 20]
     start = 0
     cite_cnt = 1
     threshold = 13.5
     cited_paper_id_to_cnt = {}
     cited_paper_list = []
+
     for sentence in sentences:
         left = final_response.find(sentence, start)
         right = left + len(sentence)
         relevant_nodes = bm25_retriever.retrieve(sentence)
+
         if len(relevant_nodes) == 0 or len(sentence.strip()) < 20:
             start = right
             continue
+
         if len(relevant_nodes) == 1 or relevant_nodes[0].node.metadata['id'] == relevant_nodes[1].node.metadata['id']:
-            paper1 = relevant_nodes[0]
-            paper1_id = paper1.node.metadata['id']
-            paper1_title = paper1.node.metadata['title']
-            if paper1.score > threshold:
-                if paper1_id not in cited_paper_id_to_cnt:
-                    cited_paper_id_to_cnt[paper1_id] = cite_cnt
-                    cited_paper_list.append((paper1_id, paper1_title))
-                    cite_cnt += 1
-                paper1_cite_cnt = cited_paper_id_to_cnt[paper1_id]
-                cite_str = f"[[{paper1_cite_cnt}]](https://arxiv.org/abs/{paper1_id})"
-                final_response = final_response[:right - 1] + cite_str + final_response[right - 1:]
+            paper = relevant_nodes[0]
+            if paper.score > threshold:
+                cite_cnt, cite_cnt = process_citation(paper, cited_paper_id_to_cnt, cited_paper_list, cite_cnt)
+                cite_str = create_cite_string(paper.node.metadata['id'], cite_cnt - 1)
```

Comment:
Two variables with the same name. If the paper ID is already in the dictionary, the numbering in cite_str will be incorrect.

Reply:
The rest of the code is fine. Thank you for your suggested improvement!

Reply:
hey sure I will fix it dw